x/ref/lib/vdl/codegen/json/const.go

```diff
@@ -0,0 +1,116 @@
+// Package json implements JSON generation for VDL const values.
+package json
+
+import (
+	"fmt"
+	"strconv"
+
+	"v.io/v23/vdl"
+	"v.io/x/ref/lib/vdl/codegen"
+	"v.io/x/ref/lib/vdl/vdlutil"
+	"v.io/x/ref/lib/vdl/codegen/vdlgen"
+)
+
+// Const returns a JSON representation of a value.
+//
+// TODO(razvanm): use this function to replace most (all?) of the untypedConst
+// function from v.io/x/ref/lib/vdl/codegen/javascript.
+func Const(v *vdl.Value, pkgPath string, imports codegen.Imports) string {
+	switch v.Kind() {
+	case vdl.Bool:
+		if v.Bool() {
+			return "true"
+		}
+		return "false"
+	case vdl.Byte, vdl.Uint16, vdl.Uint32:
+		return strconv.FormatUint(v.Uint(), 10)
+	case vdl.Int8, vdl.Int16, vdl.Int32:
+		return strconv.FormatInt(v.Int(), 10)
+	case vdl.Uint64:
+		return fmt.Sprintf(`"%d"`, v.Uint())
+	case vdl.Int64:
+		return fmt.Sprintf(`"%d"`, v.Int())
+	case vdl.Float32, vdl.Float64:
+		return strconv.FormatFloat(v.Float(), 'g', -1, bitlen(v.Kind()))
+	case vdl.String:
+		return strconv.Quote(v.RawString())
+	case vdl.Any:
+		if elem := v.Elem(); elem != nil {
+			return Const(elem, pkgPath, imports)
+		}
+		return "null"
+	case vdl.Optional:
+		if elem := v.Elem(); elem != nil {
+			return Const(elem, pkgPath, imports)
+		}
+		return "null"
+	case vdl.Enum:
+		return fmt.Sprintf(`"%s"`, v.EnumLabel())
+	case vdl.Array, vdl.List:
+		result := "["
+		for ix := 0; ix < v.Len(); ix++ {
+			if ix > 0 {
+				result += ","
+			}
+			val := Const(v.Index(ix), pkgPath, imports)
+			result += val
+		}
+		result += "]"
+		return result
+	case vdl.Set:
+		result := "["
+		for ix, key := range vdl.SortValuesAsString(v.Keys()) {
+			if ix > 0 {
+				result += ","
+			}
+			result += Const(key, pkgPath, imports)
+		}
+		result += "]"
+		return result
+	case vdl.Map:
+		result := "{"
+		for i, key := range vdl.SortValuesAsString(v.Keys()) {
+			if i > 0 {
+				result += ","
+			}
+			result += fmt.Sprintf(`%s: %s`, quote(Const(key, pkgPath, imports)), Const(v.MapIndex(key), pkgPath, imports))
+		}
+		result += "}"
+		return result
+	case vdl.Struct:
+		result := "{"
+		t := v.Type()
+		for ix := 0; ix < t.NumField(); ix++ {
+			if ix > 0 {
+				result += ","
+			}
+			result += fmt.Sprintf(`"%s": %s`, vdlutil.FirstRuneToLower(t.Field(ix).Name), Const(v.StructField(ix), pkgPath, imports))
+		}
+		return result + "}"
+	case vdl.Union:
+		ix, innerVal := v.UnionField()
+		return fmt.Sprintf("{ %q: %v }", vdlutil.FirstRuneToLower(v.Type().Field(ix).Name), Const(innerVal, pkgPath, imports))
+	case vdl.TypeObject:
+		// TODO(razvanm): check it this is correct.
+		return vdlgen.Type(v.TypeObject(), pkgPath, imports)
+	default:
+		panic(fmt.Errorf("vdl: untypedConst unhandled type %v %v", v.Kind(), v.Type()))
+	}
+}
+
+func bitlen(kind vdl.Kind) int {
+	switch kind {
+	case vdl.Float32:
+		return 32
+	case vdl.Float64:
+		return 64
+	}
+	panic(fmt.Errorf("vdl: bitLen unhandled kind %v", kind))
+}
+
+func quote(s string) string {
```

Comment:
This function is weird.  First off, it's used to quote map keys, which can be a variety of data types (I forget the exact VDL rules).  So it seems kinda dangerous, if the key itself has embedded double-quotes.

The implementation is also wrong.  E.g. in this implementation,
```
quote("a") == "a"
```
whereas I think the intention is
```
quote("a") == `"a"`
```

Reply:
This was something I did :( It was a quick change to quote map keys for JSON outputs. The map keys can indeed be a variety of data types, so perhaps the better solution would've been to panic in the case the key contains a non-string type?